File: docs/references/strategy-notes/strategy1.js
const fs = require('fs');
const path = require('path');

function parseArgs(argv) {
  const args = { input: './Data/idx_data.csv', output: './Position/New_Consecutive_Red_Candle_Daily_JS', fixedMax: false, maxRed: 3 };
  for (let i = 2; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === '--input' && argv[i + 1]) args.input = argv[++i];
    else if (a === '--output' && argv[i + 1]) args.output = argv[++i];
    else if (a === '--fixed-max') args.fixedMax = true;
    else if (a === '--max-red' && argv[i + 1]) args.maxRed = Number(argv[++i]);
  }
  return args;
}

function getParams(ticker, fixedMax, manualMax) {
  let maxRed;
  if (fixedMax) maxRed = manualMax;
  else if (ticker === 'NIFTY 50') maxRed = 7;
  else if (ticker === 'NIFTYBANK') maxRed = 5;
  else maxRed = 6;

  return {
    maxRed,
    upcloseCount: ticker !== 'NIFTY 50' ? 2 : 3,
    atrPeriod: 5,
    atrFactor: ticker !== 'NIFTYMIDCAP100' ? 2 : 4,
  };
}

function parseCsv(filePath) {
  const raw = fs.readFileSync(filePath, 'utf8').trim();
  const lines = raw.split(/\r?\n/);
  const header = lines[0].split(',');
  const idx = Object.fromEntries(header.map((h, i) => [h, i]));

  const rows = [];
  for (let i = 1; i < lines.length; i += 1) {
    const c = lines[i].split(',');
    if (c.length < 6) continue;
    rows.push({
      Ticker: c[idx.Ticker],
      Date: c[idx.Date],
      Open: Number(c[idx.Open]),
      High: Number(c[idx.High]),
      Low: Number(c[idx.Low]),
      Close: Number(c[idx.Close]),
    });
  }
  return rows;
}

function dedupSort(rows) {
  rows.sort((a, b) => {
    if (a.Ticker < b.Ticker) return -1;
    if (a.Ticker > b.Ticker) return 1;
    if (a.Date < b.Date) return -1;
    if (a.Date > b.Date) return 1;
    return 0;
  });

  const out = [];
  for (let i = 0; i < rows.length; i += 1) {
    if (out.length === 0) {
      out.push(rows[i]);
      continue;
    }
    const p = out[out.length - 1];
    const r = rows[i];
    if (p.Ticker === r.Ticker && p.Date === r.Date) out[out.length - 1] = r;
    else out.push(r);
  }
  return out;
}

function groupByTicker(rows) {
  const m = new Map();
  for (const r of rows) {
    if (!m.has(r.Ticker)) m.set(r.Ticker, []);
    m.get(r.Ticker).push(r);
  }
  return m;
}

function consecutiveCounts(mask) {
  const out = new Int32Array(mask.length);
  let run = 0;
  for (let i = 0; i < mask.length; i += 1) {
    if (mask[i]) run += 1;
    else run = 0;
    out[i] = run;
  }
  return out;
}

function rollingMean(arr, period) {
  const out = new Float64Array(arr.length);
  let sum = 0;
  for (let i = 0; i < arr.length; i += 1) {
    sum += arr[i];
    if (i >= period) sum -= arr[i - period];
    out[i] = i >= period - 1 ? sum / period : NaN;
  }
  return out;
}

function searchFirstGe(csum, target, startIdx) {
  let lo = startIdx;
  let hi = csum.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (csum[mid] >= target) hi = mid;
    else lo = mid + 1;
  }
  return lo;
}

function round2(n) {
  return Math.round(n * 100) / 100;
}

function toCsvValue(v) {
  if (v === null || v === undefined || Number.isNaN(v)) return '';
  return String(v);
}

function writeTrades(filePath, trades) {
  const header = 'Ticker,Entry Date,Exit Date,Entry Price,Exit Price,% Chg\n';
  const body = trades
    .map((t) => [t.Ticker, t['Entry Date'], t['Exit Date'], t['Entry Price'], t['Exit Price'], t['% Chg']].map(toCsvValue).join(','))
    .join('\n');
  fs.writeFileSync(filePath, header + body + (body ? '\n' : ''), 'utf8');
}

function runBacktest(rows, outputDir, fixedMax, manualMaxRed) {
  fs.mkdirSync(outputDir, { recursive: true });
  const groups = groupByTicker(dedupSort(rows));

  for (const [ticker, g] of groups.entries()) {
    const p = getParams(ticker, fixedMax, manualMaxRed);
    const n = g.length;

    const dates = new Array(n);
    const open = new Float64Array(n);
    const high = new Float64Array(n);
    const low = new Float64Array(n);
    const close = new Float64Array(n);

    for (let i = 0; i < n; i += 1) {
      const r = g[i];
      dates[i] = r.Date;
      open[i] = r.Open;
      high[i] = r.High;
      low[i] = r.Low;
      close[i] = r.Close;
    }

    const redMask = new Uint8Array(n);
    for (let i = 0; i < n; i += 1) redMask[i] = open[i] > close[i] ? 1 : 0;
    const consecRed = consecutiveCounts(redMask);

    const entryIdx = [];
    for (let i = 0; i < n; i += 1) if (consecRed[i] >= p.maxRed) entryIdx.push(i);
    if (entryIdx.length === 0) continue;

    const upclose = new Uint8Array(n);
    for (let i = 1; i < n; i += 1) upclose[i] = close[i] > close[i - 1] ? 1 : 0;

    const csum = new Int32Array(n);
    let s = 0;
    for (let i = 0; i < n; i += 1) {
      s += upclose[i];
      csum[i] = s;
    }

    const hl = new Float64Array(n);
    const hpc = new Float64Array(n);
    const lpc = new Float64Array(n);
    const tr = new Float64Array(n);

    hl[0] = high[0] - low[0];
    hpc[0] = NaN;
    lpc[0] = NaN;
    tr[0] = hl[0];
    for (let i = 1; i < n; i += 1) {
      hl[i] = high[i] - low[i];
      hpc[i] = Math.abs(high[i] - close[i - 1]);
      lpc[i] = Math.abs(low[i] - close[i - 1]);
      tr[i] = Math.max(hl[i], hpc[i], lpc[i]);
    }

    const atr = rollingMean(tr, p.atrPeriod);
    const finalAtrPlot = new Float64Array(n);
    for (let i = 0; i < n; i += 1) {
      const avg = (high[i] + low[i]) / 2;
      finalAtrPlot[i] = Number.isNaN(atr[i]) ? NaN : avg - atr[i] * p.atrFactor;
    }

    const upcloseTrades = [];
    let lastExit = -1;
    for (const e of entryIdx) {
      if (e <= lastExit) continue;
      const target = csum[e] + p.upcloseCount;
      const x = searchFirstGe(csum, target, e);
      const entryPrice = round2(close[e]);

      if (x >= n) {
        upcloseTrades.push({
          Ticker: ticker,
          'Entry Date': dates[e],
          'Exit Date': '',
          'Entry Price': entryPrice,
          'Exit Price': '',
          '% Chg': '',
        });
        break;
      }

      const exitPrice = round2(close[x]);
      upcloseTrades.push({
        Ticker: ticker,
        'Entry Date': dates[e],
        'Exit Date': dates[x],
        'Entry Price': entryPrice,
        'Exit Price': exitPrice,
        '% Chg': round2((100 * (exitPrice - entryPrice)) / entryPrice),
      });
      lastExit = x;
    }

    const atrTrades = [];
    lastExit = -1;
    for (const e of entryIdx) {
      if (e <= lastExit || Number.isNaN(finalAtrPlot[e])) continue;

      let trailing = finalAtrPlot[e];
      let exitIdx = -1;
      for (let i = e + 1; i < n; i += 1) {
        if (close[i] < trailing) {
          exitIdx = i;
          break;
        }
        trailing = Math.max(trailing, finalAtrPlot[i]);
      }

      const entryPrice = round2(close[e]);
      if (exitIdx < 0) {
        atrTrades.push({
          Ticker: ticker,
          'Entry Date': dates[e],
          'Exit Date': '',
          'Entry Price': entryPrice,
          'Exit Price': '',
          '% Chg': '',
        });
        break;
      }

      const exitPrice = round2(close[exitIdx]);
      atrTrades.push({
        Ticker: ticker,
        'Entry Date': dates[e],
        'Exit Date': dates[exitIdx],
        'Entry Price': entryPrice,
        'Exit Price': exitPrice,
        '% Chg': round2((100 * (exitPrice - entryPrice)) / entryPrice),
      });
      lastExit = exitIdx;
    }

    if (upcloseTrades.length) {
      writeTrades(path.join(outputDir, `${ticker}_${p.maxRed}_Consec_RedCandle_${p.upcloseCount}_UpClose_Exit.csv`), upcloseTrades);
    }
    if (atrTrades.length) {
      writeTrades(path.join(outputDir, `${ticker}_${p.maxRed}_Consec_RedCandle_ATR(${p.atrPeriod},${p.atrFactor})Exit.csv`), atrTrades);
    }
  }
}

function main() {
  const args = parseArgs(process.argv);
  const t0 = process.hrtime.bigint();
  const rows = parseCsv(args.input);
  runBacktest(rows, args.output, args.fixedMax, args.maxRed);
  const t1 = process.hrtime.bigint();
  const elapsedMs = Number(t1 - t0) / 1e6;
  console.log(`node_backtest_elapsed_ms=${elapsedMs.toFixed(2)}`);
}

main();

File: docs/references/strategy-notes/strategy1.rs
use std::collections::BTreeMap;
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

#[derive(Clone, Debug)]
struct Row {
    ticker: String,
    date: String,
    open: f64,
    high: f64,
    low: f64,
    close: f64,
}

#[derive(Clone, Copy)]
struct Params {
    max_red: usize,
    upclose_count: i32,
    atr_period: usize,
    atr_factor: i32,
}

#[derive(Clone)]
struct Trade {
    ticker: String,
    entry_date: String,
    exit_date: Option<String>,
    entry_price: f64,
    exit_price: Option<f64>,
    pct_chg: Option<f64>,
}

fn round2(v: f64) -> f64 {
    (v * 100.0).round() / 100.0
}

fn parse_args() -> (String, String, bool, usize) {
    let mut input = "./Data/idx_data.csv".to_string();
    let mut output = "./Position/New_Consecutive_Red_Candle_Daily_RS".to_string();
    let mut fixed_max = false;
    let mut max_red = 3usize;

    let args: Vec<String> = env::args().collect();
    let mut i = 1usize;
    while i < args.len() {
        match args[i].as_str() {
            "--input" if i + 1 < args.len() => {
                input = args[i + 1].clone();
                i += 1;
            }
            "--output" if i + 1 < args.len() => {
                output = args[i + 1].clone();
                i += 1;
            }
            "--fixed-max" => {
                fixed_max = true;
            }
            "--max-red" if i + 1 < args.len() => {
                max_red = args[i + 1].parse().unwrap_or(3);
                i += 1;
            }
            _ => {}
        }
        i += 1;
    }

    (input, output, fixed_max, max_red)
}

fn get_params(ticker: &str, fixed_max: bool, manual_max: usize) -> Params {
    let max_red = if fixed_max {
        manual_max
    } else if ticker == "NIFTY 50" {
        7
    } else if ticker == "NIFTYBANK" {
        5
    } else {
        6
    };

    Params {
        max_red,
        upclose_count: if ticker != "NIFTY 50" { 2 } else { 3 },
        atr_period: 5,
        atr_factor: if ticker != "NIFTYMIDCAP100" { 2 } else { 4 },
    }
}

fn parse_csv(path: &str) -> Result<Vec<Row>, String> {
    let content = fs::read_to_string(path).map_err(|e| format!("read failed: {e}"))?;
    let mut lines = content.lines();
    let header = lines.next().ok_or("empty csv")?;
    let cols: Vec<&str> = header.split(',').collect();

    let ticker_idx = cols.iter().position(|c| *c == "Ticker").ok_or("Ticker missing")?;
    let date_idx = cols.iter().position(|c| *c == "Date").ok_or("Date missing")?;
    let open_idx = cols.iter().position(|c| *c == "Open").ok_or("Open missing")?;
    let high_idx = cols.iter().position(|c| *c == "High").ok_or("High missing")?;
    let low_idx = cols.iter().position(|c| *c == "Low").ok_or("Low missing")?;
    let close_idx = cols.iter().position(|c| *c == "Close").ok_or("Close missing")?;

    let mut rows = Vec::new();
    for line in lines {
        if line.trim().is_empty() {
            continue;
        }
        let parts: Vec<&str> = line.split(',').collect();
        if parts.len() <= close_idx {
            continue;
        }
        let open = match parts[open_idx].parse::<f64>() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let high = match parts[high_idx].parse::<f64>() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let low = match parts[low_idx].parse::<f64>() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let close = match parts[close_idx].parse::<f64>() {
            Ok(v) => v,
            Err(_) => continue,
        };

        rows.push(Row {
            ticker: parts[ticker_idx].to_string(),
            date: parts[date_idx].to_string(),
            open,
            high,
            low,
            close,
        });
    }
    Ok(rows)
}

fn sort_and_dedup(mut rows: Vec<Row>) -> Vec<Row> {
    rows.sort_by(|a, b| {
        a.ticker
            .cmp(&b.ticker)
            .then_with(|| a.date.cmp(&b.date))
    });

    let mut out: Vec<Row> = Vec::with_capacity(rows.len());
    for row in rows {
        if let Some(last) = out.last_mut() {
            if last.ticker == row.ticker && last.date == row.date {
                *last = row;
                continue;
            }
        }
        out.push(row);
    }
    out
}

fn group_by_ticker(rows: Vec<Row>) -> BTreeMap<String, Vec<Row>> {
    let mut grouped = BTreeMap::new();
    for row in rows {
        grouped
            .entry(row.ticker.clone())
            .or_insert_with(Vec::new)
            .push(row);
    }
    grouped
}

fn consecutive_counts(mask: &[bool]) -> Vec<usize> {
    let mut out = vec![0usize; mask.len()];
    let mut run = 0usize;
    for (i, v) in mask.iter().enumerate() {
        if *v {
            run += 1;
        } else {
            run = 0;
        }
        out[i] = run;
    }
    out
}

fn rolling_mean(values: &[f64], period: usize) -> Vec<Option<f64>> {
    let mut out = vec![None; values.len()];
    let mut sum = 0.0f64;
    for i in 0..values.len() {
        sum += values[i];
        if i >= period {
            sum -= values[i - period];
        }
        if i + 1 >= period {
            out[i] = Some(sum / period as f64);
        }
    }
    out
}

fn search_first_ge(values: &[i32], start: usize, target: i32) -> usize {
    let mut lo = start;
    let mut hi = values.len();
    while lo < hi {
        let mid = (lo + hi) / 2;
        if values[mid] >= target {
            hi = mid;
        } else {
            lo = mid + 1;
        }
    }
    lo
}

fn write_trades(path: &Path, trades: &[Trade]) -> Result<(), String> {
    let mut out = String::from("Ticker,Entry Date,Exit Date,Entry Price,Exit Price,% Chg\n");
    for trade in trades {
        let exit_date = trade.exit_date.clone().unwrap_or_default();
        let exit_price = trade
            .exit_price
            .map(|v| round2(v).to_string())
            .unwrap_or_default();
        let pct_chg = trade
            .pct_chg
            .map(|v| round2(v).to_string())
            .unwrap_or_default();

        out.push_str(&format!(
            "{},{},{},{},{},{}\n",
            trade.ticker,
            trade.entry_date,
            exit_date,
            round2(trade.entry_price),
            exit_price,
            pct_chg
        ));
    }
    fs::write(path, out).map_err(|e| format!("write failed: {e}"))
}

fn run_backtest(rows: Vec<Row>, output_dir: &str, fixed_max: bool, manual_max: usize) -> Result<(), String> {
    fs::create_dir_all(output_dir).map_err(|e| format!("mkdir failed: {e}"))?;
    let groups = group_by_ticker(sort_and_dedup(rows));

    for (ticker, group) in groups {
        let p = get_params(&ticker, fixed_max, manual_max);
        let n = group.len();
        if n == 0 {
            continue;
        }

        let mut dates = Vec::with_capacity(n);
        let mut open = Vec::with_capacity(n);
        let mut high = Vec::with_capacity(n);
        let mut low = Vec::with_capacity(n);
        let mut close = Vec::with_capacity(n);

        for row in &group {
            dates.push(row.date.clone());
            open.push(row.open);
            high.push(row.high);
            low.push(row.low);
            close.push(row.close);
        }

        let red_mask: Vec<bool> = open.iter().zip(close.iter()).map(|(o, c)| o > c).collect();
        let consec_red = consecutive_counts(&red_mask);
        let entry_idx: Vec<usize> = consec_red
            .iter()
            .enumerate()
            .filter_map(|(i, v)| if *v >= p.max_red { Some(i) } else { None })
            .collect();
        if entry_idx.is_empty() {
            continue;
        }

        let mut upclose = vec![0i32; n];
        for i in 1..n {
            upclose[i] = if close[i] > close[i - 1] { 1 } else { 0 };
        }

        let mut csum = vec![0i32; n];
        let mut running = 0i32;
        for i in 0..n {
            running += upclose[i];
            csum[i] = running;
        }

        let mut tr = vec![0.0f64; n];
        tr[0] = high[0] - low[0];
        for i in 1..n {
            let hl = high[i] - low[i];
            let hpc = (high[i] - close[i - 1]).abs();
            let lpc = (low[i] - close[i - 1]).abs();
            tr[i] = hl.max(hpc).max(lpc);
        }

        let atr = rolling_mean(&tr, p.atr_period);
        let mut final_atr_plot = vec![None; n];
        for i in 0..n {
            if let Some(atr_val) = atr[i] {
                let avg = (high[i] + low[i]) / 2.0;
                final_atr_plot[i] = Some(avg - atr_val * p.atr_factor as f64);
            }
        }

        let mut upclose_trades = Vec::new();
        let mut last_exit = None::<usize>;
        for &e in &entry_idx {
            if last_exit.is_some() && e <= last_exit.unwrap() {
                continue;
            }
            let target = csum[e] + p.upclose_count;
            let x = search_first_ge(&csum, e, target);
            let entry_price = round2(close[e]);
            if x >= n {
                upclose_trades.push(Trade {
                    ticker: ticker.clone(),
                    entry_date: dates[e].clone(),
                    exit_date: None,
                    entry_price,
                    exit_price: None,
                    pct_chg: None,
                });
                break;
            }
            let exit_price = round2(close[x]);
            upclose_trades.push(Trade {
                ticker: ticker.clone(),
                entry_date: dates[e].clone(),
                exit_date: Some(dates[x].clone()),
                entry_price,
                exit_price: Some(exit_price),
                pct_chg: Some(round2(100.0 * (exit_price - entry_price) / entry_price)),
            });
            last_exit = Some(x);
        }

        let mut atr_trades = Vec::new();
        last_exit = None;
        for &e in &entry_idx {
            if last_exit.is_some() && e <= last_exit.unwrap() {
                continue;
            }
            let mut trailing = match final_atr_plot[e] {
                Some(v) => v,
                None => continue,
            };
            let mut exit_idx = None::<usize>;
            for i in (e + 1)..n {
                if close[i] < trailing {
                    exit_idx = Some(i);
                    break;
                }
                if let Some(v) = final_atr_plot[i] {
                    if v > trailing {
                        trailing = v;
                    }
                }
            }

            let entry_price = round2(close[e]);
            if let Some(x) = exit_idx {
                let exit_price = round2(close[x]);
                atr_trades.push(Trade {
                    ticker: ticker.clone(),
                    entry_date: dates[e].clone(),
                    exit_date: Some(dates[x].clone()),
                    entry_price,
                    exit_price: Some(exit_price),
                    pct_chg: Some(round2(100.0 * (exit_price - entry_price) / entry_price)),
                });
                last_exit = Some(x);
            } else {
                atr_trades.push(Trade {
                    ticker: ticker.clone(),
                    entry_date: dates[e].clone(),
                    exit_date: None,
                    entry_price,
                    exit_price: None,
                    pct_chg: None,
                });
                break;
            }
        }

        if !upclose_trades.is_empty() {
            let file = Path::new(output_dir).join(format!(
                "{}_{}_Consec_RedCandle_{}_UpClose_Exit.csv",
                ticker, p.max_red, p.upclose_count
            ));
            write_trades(&file, &upclose_trades)?;
        }

        if !atr_trades.is_empty() {
            let file = Path::new(output_dir).join(format!(
                "{}_{}_Consec_RedCandle_ATR({},{})Exit.csv",
                ticker, p.max_red, p.atr_period, p.atr_factor
            ));
            write_trades(&file, &atr_trades)?;
        }
    }

    Ok(())
}

fn main() {
    let (input, output, fixed_max, max_red) = parse_args();
    let started = Instant::now();

    let rows = match parse_csv(&input) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("{e}");
            std::process::exit(1);
        }
    };

    if let Err(e) = run_backtest(rows, &output, fixed_max, max_red) {
        eprintln!("{e}");
        std::process::exit(1);
    }

    println!("rust_backtest_elapsed_ms={:.2}", started.elapsed().as_secs_f64() * 1000.0);
}
